File: C++ solutions/68. Text Justification.cpp
class Solution {
public:
    vector<string> fullJustify(vector<string>& words, int maxWidth) {
        vector<string> ans;
        int cnt = 0;
        vector<string> currentLine;
        
        for (int i = 0; i < words.size(); i++) {
            string s = words[i];
            int len = s.size();
            
            if (cnt + len + currentLine.size() > maxWidth) {
                int spacesToAdd = maxWidth - cnt;
                if (currentLine.size() == 1) {
                    ans.push_back(currentLine[0] + string(spacesToAdd, ' '));
                } else {
                    int baseSpaces = spacesToAdd / (currentLine.size() - 1);
                    int extraSpaces = spacesToAdd % (currentLine.size() - 1);
                    
                    string line = currentLine[0];
                    for (int j = 1; j < currentLine.size(); j++) {
                        line += string(baseSpaces, ' ');
                        if (j <= extraSpaces) {
                            line += ' ';
                        }
                        line += currentLine[j];
                    }
                    ans.push_back(line);
                }
                currentLine.clear();
                cnt = 0;
            }
            
            currentLine.push_back(s);
            cnt += len;
        }
        
       
        string lastLine = currentLine[0];
        for (int i = 1; i < currentLine.size(); i++) {
            lastLine += ' ' + currentLine[i];
        }
        lastLine += string(maxWidth - lastLine.size(), ' ');
        ans.push_back(lastLine);
        
        return ans;
    }
};

File: JavaScript solutions/68. Text Justification.js
/**
 * @param {string[]} words
 * @param {number} maxWidth
 * @return {string[]}
 */
var fullJustify = function (words, maxWidth) {
  let ans = [];
  let cnt = 0;
  let currentLine = [];

  for (let i = 0; i < words.length; i++) {
    let s = words[i];
    let len = s.length;

    if (cnt + len + currentLine.length > maxWidth) {
      let spacesToAdd = maxWidth - cnt;
      if (currentLine.length === 1) {
        ans.push(currentLine[0] + " ".repeat(spacesToAdd));
      } else {
        let baseSpaces = Math.floor(spacesToAdd / (currentLine.length - 1));
        let extraSpaces = spacesToAdd % (currentLine.length - 1);

        let line = currentLine[0];
        for (let j = 1; j < currentLine.length; j++) {
          line += " ".repeat(baseSpaces);
          if (j <= extraSpaces) {
            line += " ";
          }
          line += currentLine[j];
        }
        ans.push(line);
      }
      currentLine = [];
      cnt = 0;
    }

    currentLine.push(s);
    cnt += len;
  }

  let lastLine = currentLine[0];
  for (let i = 1; i < currentLine.length; i++) {
    lastLine += " " + currentLine[i];
  }
  lastLine += " ".repeat(maxWidth - lastLine.length);
  ans.push(lastLine);

  return ans;
};
